lib/web_ui/lib/src/engine/canvaskit/image.dart

```diff
@@ -4,24 +4,193 @@
+  // ImageDecoder does not detect image type automatically. It requires us to
+  // tell it what the image type is.
+  final String contentType = tryDetectContentType(list, 'encoded image bytes');
+
+  if (browserSupportsImageDecoder) {
+    codec = await CkBrowserImageDecoder.create(
       data: list,
+      contentType: contentType,
       debugSource: 'encoded image bytes',
     );
   } else {
-    return CkAnimatedImage.decodeFromBytes(list, 'encoded image bytes', targetWidth: targetWidth, targetHeight: targetHeight);
+    // TODO(harryterkelsen): If the image is animated, then use Skia to decode.
+    // This is currently too conservative, assuming all GIF and WEBP images are
+    // animated. We should detect if they are actually animated by reading the
+    // image headers, https://github.com/flutter/flutter/issues/151911.
+    if (contentType == 'image/gif' || contentType == 'image/webp') {
+      codec = CkAnimatedImage.decodeFromBytes(list, 'encoded image bytes',
+          targetWidth: targetWidth, targetHeight: targetHeight);
+    } else {
+      final DomBlob blob = createDomBlob(<ByteBuffer>[list.buffer]);
+      codec = await decodeBlobToCkImage(blob);
+    }
+  }
+  return CkResizingCodec(
+    codec,
+    targetWidth: targetWidth,
+    targetHeight: targetHeight,
+    allowUpscaling: allowUpscaling,
+  );
+}
+
+/// A resizing codec which uses an HTML <canvas> element to scale the image if
+/// it is backed by an HTML Image element.
+class CkResizingCodec extends ResizingCodec {
+  CkResizingCodec(
+    super.delegate, {
+    super.targetWidth,
+    super.targetHeight,
+    super.allowUpscaling,
+  });
+
+  @override
+  ui.Image scaleImage(
+    ui.Image image, {
+    int? targetWidth,
+    int? targetHeight,
+    bool allowUpscaling = true,
+  }) {
+    final CkImage ckImage = image as CkImage;
+    if (ckImage.imageSource == null) {
+      return scaleImageIfNeeded(
+        image,
+        targetWidth: targetWidth,
+        targetHeight: targetHeight,
+        allowUpscaling: allowUpscaling,
+      );
+    } else {
+      return _scaleImageUsingDomCanvas(
+        ckImage,
+        targetWidth: targetWidth,
+        targetHeight: targetHeight,
+        allowUpscaling: allowUpscaling,
+      );
+    }
+  }
+
+  CkImage _scaleImageUsingDomCanvas(
+    CkImage image, {
+    int? targetWidth,
+    int? targetHeight,
+    bool allowUpscaling = true,
+  }) {
+    assert(image.imageSource != null);
+    final int width = image.width;
+    final int height = image.height;
+    final BitmapSize? scaledSize =
+        scaledImageSize(width, height, targetWidth, targetHeight);
+    if (scaledSize == null) {
+      return image;
+    }
+    if (!allowUpscaling &&
+        (scaledSize.width > width || scaledSize.height > height)) {
+      return image;
+    }
+
+    final int scaledWidth = scaledSize.width;
+    final int scaledHeight = scaledSize.height;
+
+    final DomOffscreenCanvas offscreenCanvas = createDomOffscreenCanvas(
```

Comment:
I think you can avoid creating an offscreen canvas and so on here by just using `createImageBitmap` instead. https://developer.mozilla.org/en-US/docs/Web/API/createImageBitmap It has `resizeWidth` and `resizeHeight` options which will to the rescaling for you.

Reply:
Harry pointed out that `createImageBitmap` is async, which makes this much less clean. So we should leave it as is.